subsys/bluetooth/mesh/proxy_srv.c

```diff
@@ -680,155 +658,249 @@ static bool sub_count_cb(struct bt_mesh_subnet *sub, void *cb_data)
-		priv_net_id_adv(sub, remaining);
+		LOG_DBG("Advertising Private Network ID timed out after solicitation");
+	}
+#endif
+}
 
-		if (!sub->priv_net_id_sent) {
-			sub->priv_net_id_sent = now;
+static bool adv_sub_check(struct bt_mesh_subnet *sub, struct proxy_adv_req *req)
+{
+	if (sub->net_idx == BT_MESH_KEY_UNUSED) {
+		return false;
+	}
+
+	if (sub->node_id == BT_MESH_NODE_IDENTITY_RUNNING) {
+		bool priv_node_id = false;
+
+		req->duration = NODE_ID_TIMEOUT - (k_uptime_get_32() - sub->node_id_start);
+#if defined(CONFIG_BT_MESH_PRIV_BEACONS)
+		priv_node_id = sub->priv_beacon_ctx.node_id;
+#endif
+		req->evt = priv_node_id ? PRIV_NODE_ID : NODE_ID;
+
+		/** Covers corner case where we have subsecond duration.
+		 *  Advertiser does not handle subsecond duration well
+		 *  since it will cause spamming of advertsiments. To
+		 *  prevent this we reset the Node Identity adv state
+		 *  here, but let it advertise for one additional second.
+		 */
+		if (req->duration < MSEC_PER_SEC) {
+			bt_mesh_proxy_identity_stop(sub);
+			LOG_DBG("Node ID stopped for subnet %d", sub->net_idx);
+		}
+
+		return true;
+	}
+
+#if defined(CONFIG_BT_MESH_OD_PRIV_PROXY_SRV)
+	if (bt_mesh_od_priv_proxy_get() > 0 && sub->solicited) {
+		int32_t timeout = MSEC_PER_SEC * (int32_t)bt_mesh_od_priv_proxy_get();
+
+		req->evt = OD_PRIV_NODE_ID;
+		req->duration = !sub->priv_net_id_sent
+					? timeout
+					: timeout - (k_uptime_get_32() - sub->priv_net_id_sent);
+
+		/** Covers corner case where we have subsecond duration.
+		 *  Advertiser does not handle subsecond duration well
+		 *  since it will cause spamming of advertsiments. To
+		 *  prevent this we reset the OD Priv Proxy adv state
+		 *  here, but let it advertise for one additional second.
+		 */
+		if (req->duration < MSEC_PER_SEC) {
+			sub->priv_net_id_sent = 0;
+			sub->solicited = false;
+			LOG_DBG("Advertising Private Network ID timed out  after solicitation");
 		}
+
+		return true;
 	}
-}
 #endif
 
-static int gatt_proxy_advertise(struct bt_mesh_subnet *sub)
+	if (bt_mesh_priv_gatt_proxy_get() == BT_MESH_FEATURE_ENABLED) {
+		req->evt = PRIV_NET_ID;
+		req->duration = PROXY_RANDOM_UPDATE_INTERVAL;
+		return true;
+	}
+
+	if (bt_mesh_gatt_proxy_get() == BT_MESH_FEATURE_ENABLED) {
+		req->evt = NET_ID;
+		req->duration = SYS_FOREVER_MS;
+		return true;
+	}
+
+	return false;
+}
+
+static struct bt_mesh_subnet *adv_sub_get(struct bt_mesh_subnet *sub_start,
+					  struct proxy_adv_req *req, bool specific)
+{
+	struct bt_mesh_subnet *sub_temp = sub_start;
+
+	req->evt = NO_ADV;
+
+	do {
+		if (adv_sub_check(sub_temp, req)) {
+			return sub_temp;
+		}
+
+		sub_temp = bt_mesh_subnet_next(sub_temp);
+	} while (sub_temp != sub_start && !specific);
+
+	return NULL;
+}
+
+static int gatt_proxy_advertise(void)
 {
-	int32_t remaining = SYS_FOREVER_MS;
-	int subnet_count;
-	int err = -EBUSY;
-	bool planned = false;
+	int err;
+
+	int32_t max_adv_duration;
+	int active_sub_cnt;
+	static int32_t prev_start;
```

Comment:
There are 4 static variables those characterize the state of the previous subnetwork advertising.
I recommend creating something like:
```
static struct {
    int32_t prev_start;
    struct bt_mesh_subnet *prev_sub;
    struct proxy_adv_req prev_req;
    bool ignore_previous;
} prev_sub_adv;
```
and keep it outside the function. It will help to keep this information in one place and make it available from for debugging\logging in other places.